src/main/BlockPool.cpp

```diff
@@ -0,0 +1,863 @@
+/*
+ *  R : A Computer Language for Statistical Data Analysis
+ *  Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
+ *  Copyright (C) 1998-2007   The R Development Core Team.
+ *  Copyright (C) 2008-2014  Andrew R. Runnalls.
+ *  Copyright (C) 2014 and onwards the Rho Project Authors.
+ *
+ *  Rho is not part of the R project, and bugs and other issues should
+ *  not be reported via r-bugs or other R project channels; instead refer
+ *  to the Rho website.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#include "rho/BlockPool.hpp"
+
+#include <assert.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <limits>
+#include <map>
+
+typedef std::uint64_t u64;
+typedef std::uint32_t u32;
+using std::function;
+
+#ifdef ALLOCATION_CHECK
+typedef std::map<void*, void*> allocation_map;
+static allocation_map allocations;
+
+static bool iterating = false;
+
+static void add_to_allocation_map(void* allocation, size_t size);
+static void remove_from_allocation_map(void* allocation);
+static void* lookup_in_allocation_map(void* tentative_pointer);
+#endif
+
+static void allocerr(const char* message) {
+  fprintf(stderr, "ERROR: %s\n", message);
+  abort();
+}
+
+using std::size_t;
+
+// Tracking heap bounds for fast rejection in pointer lookup.
+void* heap_start = reinterpret_cast<void*>(UINTPTR_MAX);
```

Comment:
Variables and function that aren't exported should be declared 'static' or in an anonymous namespace.